### ISSUE TITLE
Handle Error

### ISSUE THREAD

**Issue by chrlsbyrndlcrz** [wnabil/ngx-drupal7-services#7]
I get the error and displayed in "console.log" but i want to display in toast

Please help me for this
Thanks

**Comment by sannminwin** [comment]
Hi @wnabil ,
Thanks for this great module. It works out of the box.
I have question about error handling.
I use login function like this.
I am expection error object form "err" .
But I can not access "err" value.  
What would be best way to access error ?


this.userService.login(f.value)
    .subscribe((data) => {
      console.log('datata from login', data);
      this.cookieProviderService.setCookie('currentUser', data.user, true);
       this.router.navigate(['/home']);
    },
    (err) => {
      console.log(err);
      if (err.status == '401') {
        this.invalidLogin = true;
        this.router.navigate(['/login']);
      }
      if (err.status == '406') {
        this.blockedAccount = true;
        this.router.navigate(['/login']);
      }
    },
    () => {
      console.log('loaded login');
    });

**Comment by wnabil** [comment]
Hi
If u set the allowOffline tag, your the observable error will never be triggered.
i suggest you to remove the allowOffline and made an interceptor to angular HttpClient.
this is an example of the implimintation:
https://alligator.io/angular/httpclient-interceptors/

**Comment by centogram** [comment]
Is there a way to both allow for the offline access flag to be true and still trigger the observables?

Our Ionic app has some offline features in the event of the drupal instance being disconnected but we still wish to capture the observable responses on an individual basis.

**Comment by wnabil** [comment]
@centogram Yes, just don't use the offline mode in the package config and use the interceptors to handle your request and fetch the data from the cache.
here is an example code
`
/**
   * intercept the http requests and add loading progress and handle error
   * @param req the current request
   * @param next the next http request handler
   * @return observable http event
   * @see https://angular.io/api/common/http/HttpInterceptor
   */
  intercept(dupReq : HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    // create new observable
    const source: Observable<HttpEvent<any>> = Observable.create((observer: Observer<HttpEvent<any>>) => {
      // if the app is already offline and cordova available
      if (!this.customService.connected && this.deviceDetectService.isCordova) {
        // get the cached event
        this.getFromCache(observer, resourceName);
        return;
      }
      // if the user is online or cordova not available, send the request as normal
      next
        .handle(dupReq)
        .pipe(
          tap(
            (event: HttpEvent<any>) => {
              // set connected flag
              this.customService.connected = true;
              // in case of response
              if (event instanceof HttpResponse) {
                // send the event and finish the observer
                observer.next(event);
                observer.complete();
                // if cordova available cache the content
                if (this.deviceDetectService.isCordova) {
                  const contentToCache = <any>{ ...event };
                    // cache the file
                    this.customService.cacheFile(contentToCache);
                }
              }
            },
            (err: any) => {
              // in case of error and cordova available
              if (this.deviceDetectService.isCordova && err.status === 0) {
                // temp fix for first request, wait for cordova device ready
                setTimeout(() => {
                  // get from cache
                  this.getFromCache(observer, resourceName);
                }, 1000);
                return;
              }
              // if cordova not available let service worker do the job
              observer.error(err);
              observer.complete();
            }
          )
        )
        .subscribe();
    });
    return source;
  }
`

**Comment by centogram** [comment]
@wnabil Thank you for the nice followup on this.  Would you mind sharing your getFromCache method and your customService class as well?

**Comment by wnabil** [comment]
customService class is just a normal inherited class from mainService to retrieve the custom resources that implemented at the backend, Sorry i can not share the hole code because its a private project but if you want check any of the services in the package just like NodeService or ViewService
getFromCache is a simple method to get file from cordova file plugin
```javascript
/**
   * get single resource from cache and sent the data
   * @param observer the event observer to send the data with it
   * @param resourceName resource or file name
   */
  getFromCache(observer: Observer<any>, resourceName: string) {
    // get from custom service by resource name - cordova file plugin
    this.customService.getCachedFile(resourceName + '.json').subscribe(
      (data: any) => {
        // this line cost about 4 hours of warking, holy *
        const event = new HttpResponse(data);
        observer.next(event);
        observer.complete();
        this.afterResponse();
      },
      err => {
        // in case of file not found, display no internet message and navigate back
        this.navDataService.notificationMessage = this.notificationMessage;
        this.navDataService.navigateBack();
        this.afterResponse();
      }
    );
  }
```

**Comment by centogram** [comment]
@wnabil Thank you for the suggestions.  This will work!
